Updating to latest LLVM syntax changes.

sunfishcode opened the pull request:
With r252910 landing in LLVM, there were some syntax changes to the .s syntax, because we're now using MC for more things. This is a Good Thing in general, because it means the .s syntax is more like a regular assembly syntax (and there are other longer-term benefits of using MC besides). This patch updates wasmate.py accordingly. Major changes:
 - MC doesn't prefix symbols with '$'. This is now moved into wasmate.py.
 - push/pop are now prefixed with a $ to disambiguate them.
 - it inspired the addition of a new pseudo-operand, $discard, which is like $push in that it consumes a value, but does not push anything on the stack.
 - import symbols are no longer quoted; wasmate.py now does this.

A particularly fun part about this change in LLVM is that it eliminates stuff like this:
       get_local push, 0
       set_local 1, pop
for every argument. (And there's more than can be done in that direction.)

sunfishcode commented on prototype-wasmate/wasmate.py:
```diff
@@ -294,7 +298,8 @@ def sexprify(command, args):
         s += ' '
-    s += ' '.join([resolve_label(arg) for arg in args if arg != 'push'])
+    s += ' '.join([resolve_label(arg) for arg in args
+                   if arg != '$push' and arg != '$discard'])
```
Having $push and $discard as arguments in the LLVM output makes sense because LLVM customarily represents the "output" of an instruction as an operand, typically the output register. That said, I agree that the current syntax isn't great; I expect it to evolve significantly.

ncbray replied:
The blocker here is $discard.  What does it buy?  If we discard $discard, what is lost?

$push is an aside that does not need to be addressed in this issue.  Making "push" an outer expression would be much more natural for the rest of the pipeline.  Optional out params as the first argument is neither fish nor fowl.  If LLVM's preferences really factored into this, we'd be doing CFGs. :/

And, btw, this CL blocks be to hell and back in terms of rolling deps in wasm-e2e and landing my C raytracer example.  In the future, if you could make two-sided changes be less disruptive (say: liberalizing wasmate before landing LLVM changes) I'd appreciate it.  Assume I'll need to roll in a new version of LLVM and/or wasmate at any point.

sunfishcode replied:
There are some major changes in the syntax coming hopefully in the next few days (the abovementioned evolution). It will be very inconvenient to make wasmate.py support both old and new syntax at the same time. I consider it a higher priority to move the LLVM codegen path forward than to keep wasmate working with old versions of LLVM right now.

ncbray commented on prototype-wasmate/test/cfg-stackify.s:
```diff
@@ -370,35 +327,36 @@ func_end10:
-	br $BB11_3
+	loop BB11_3
+	block BB11_5
+	block BB11_2
```
While trying to rebase onto master, I noticed this block was malformed (termination label is above it).

The termination label for the loop is also missing.  I added a few of these labels by hand, assuming this was old code.  But apparently it's being generated from somewhere?

ncbray commented on prototype-wasmate/test/cfg-stackify.s:
```diff
@@ -370,35 +327,36 @@ func_end10:
+	block BB11_2
+	i32.eq $push, (get_local 3), (get_local 1)
+	set_local 2, $pop
+	br_if BB11_4, (get_local 2)
```
This statement is also malformed because there is no loop or block for BB11_4.

sunfishcode commented:
Updated to handle what LLVM trunk is now emitting. This includes the changes to $push/$pop/$discard, operand ordering, reg coloring, reg stackifying, cfg stackification, and other things. Several bugs in LLVM that wasmate.py was previously working around have now been fixed.

I also incorporated your review comments above.

There were some hairy merge conflicts; I've done by best to preserve the changes, but I essentially had to redo 0e886895599a54fb1290ba3ef96155b97b8e5b92, while 9edc0c3a371c166be6d5c4c1db24799fa27d7960 and 0483ab8d34a8a5ac6e023bbbfc5e0177518e78fa were fairly straightforward but still took some work; hopefully I preserved the meaning.

Also, many of the test/*.s files are generated (by scripts/llvm-to-s.py), so manual edits to them get dropped.

jfbastien commented on prototype-wasmate/wasmate.py:
```diff
@@ -72,8 +72,10 @@ def error(message, line_number=None):
-        return str(data_labels[arg[1:]])
+    if data_labels.has_key(arg):
+        return str(data_labels[arg])
+    if arg[0].isalpha() and arg != 'infinity' and arg != 'nan':
```
Do we also output `nan(...)` from LLVM? Maybe `infinity` and `nan` should be split off into their own `is_special` helper or something.

sunfishcode replied:
NaN payloads aren't yet implemented in the LLVM WebAssembly backend. But yes, once we do that, we'll likely need to do something special here.

jfbastien replied:
Ah OK, that would just fail to parse anyways so it won't be a silent wasmate breakage. Seems fine to punt.

jfbastien commented on prototype-wasmate/wasmate.py:
```diff
@@ -262,7 +264,9 @@ def __init__(self):
     def end_pass(self):
         for i in self.imports:
-            out.write_line('(import ' + i + ')')
+            parts = i.split(None, 3)
```
Could you add a comment on what the 3 parts are (and the fourth optional one)? It'll make this easier to read.

jfbastien commented:
Two nits then lgtm since this moves us forward and we can make this pretty later when there's less churn.

> Also, many of the test/*.s files are generated (by scripts/llvm-to-s.py), so manual edits to them get dropped.

Could you clarify: did you regenerate the test files from the LLVM ones? Or were there manual edits in the previous version?

sunfishcode commented:
> Could you clarify: did you regenerate the test files from the LLVM ones? Or were there manual edits in the previous version?

Yes. scripts/llvm-to-s.py is a script which auto-generates wasmate tests from LLVM WebAssembly codegen tests. This commit includes an update to the latest tests using this script. There were some prior manual edits which have been overwritten, but fortunately I believe those changes are no longer needed, due to various bugs being fixed.

jfbastien commented:
lgtm
